In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import os
import pickle

import numpy as np
import pandas as pd

import settings as conf
import metadata

# Load S-MultiXcan results

In [3]:
smultixcan_pvalues_file = os.path.join(conf.GENE_ASSOC_DIR, f'smultixcan-mashr-pvalues.pkl.xz')
display(smultixcan_pvalues_file)

'/mnt/phenomexcan_base/gene_assoc/smultixcan-mashr-pvalues.pkl.xz'

In [4]:
smultixcan_gene_associations = pd.read_pickle(smultixcan_pvalues_file)

In [5]:
smultixcan_gene_associations.shape

(22515, 4091)

In [6]:
smultixcan_gene_associations.head(5)

20096_1-Size_of_red_wine_glass_drunk_small_125ml  \
gene_name                                                           
ENSG00000000419                                          0.865429   
ENSG00000000457                                          0.174192   
ENSG00000000460                                          0.879969   
ENSG00000000938                                          0.192670   
ENSG00000000971                                          0.180632   

                 2345-Ever_had_bowel_cancer_screening  \
gene_name                                               
ENSG00000000419                              0.918314   
ENSG00000000457                              0.064765   
ENSG00000000460                              0.240715   
ENSG00000000938                              0.400054   
ENSG00000000971                              0.793060   

                 N49-Diagnoses_main_ICD10_N49_Inflammatory_disorders_of_male_genital_organs_not_elsewhere_classified  \
gene_name                                                                                                              
ENSG00000000419                                           0.810683                                                     
ENSG00000000457                                           0.889194                                                     
ENSG00000000460                                           0.238228                                                     
ENSG00000000938                                           0.114353                                                     
ENSG00000000971                                           0.490585                                                     

                 100011_raw-Iron  \
gene_name                          
ENSG00000000419         0.374671   
ENSG00000000457         0.896938   
ENSG00000000460         0.567555   
ENSG00000000938         0.470700   
ENSG00000000971         0.088752   

                 5221-Index_of_best_refractometry_result_right  \
gene_name                                                        
ENSG00000000419                                       0.189032   
ENSG00000000457                                       0.448596   
ENSG00000000460                                       0.921320   
ENSG00000000938                                       0.889202   
ENSG00000000971                                       0.744531   

                 20003_1141150624-Treatmentmedication_code_zomig_25mg_tablet  \
gene_name                                                                      
ENSG00000000419                                           0.140981             
ENSG00000000457                                           0.269602             
ENSG00000000460                                           0.825036             
ENSG00000000938                                           0.000011             
ENSG00000000971                                           0.949639             

                 S69-Diagnoses_main_ICD10_S69_Other_and_unspecified_injuries_of_wrist_and_hand  \
gene_name                                                                                        
ENSG00000000419                                           0.467741                               
ENSG00000000457                                           0.540261                               
ENSG00000000460                                           0.782230                               
ENSG00000000938                                           0.899764                               
ENSG00000000971                                           0.253817                               

                 20024_1136-Job_code_deduced_Information_and_communication_technology_managers  \
gene_name                                                                                        
ENSG00000000419                                           0.129427                               
ENSG00000000457                   

# Load fastENLOC results

In [7]:
fastenloc_rcp_file = os.path.join(conf.GENE_ASSOC_DIR, f'fastenloc-torus-rcp.pkl.xz')
display(fastenloc_rcp_file)

'/mnt/phenomexcan_base/gene_assoc/fastenloc-torus-rcp.pkl.xz'

In [8]:
fastenloc_gene_associations = pd.read_pickle(fastenloc_rcp_file)

In [9]:
fastenloc_gene_associations.shape

(37967, 4091)

In [10]:
fastenloc_gene_associations.head(5)

20096_1-Size_of_red_wine_glass_drunk_small_125ml  \
gene_id                                                             
ENSG00000000419                                      2.173200e-07   
ENSG00000000457                                      1.343900e-06   
ENSG00000000460                                      8.757750e-06   
ENSG00000000938                                      7.197000e-08   
ENSG00000000971                                      2.720000e-07   

                 2345-Ever_had_bowel_cancer_screening  \
gene_id                                                 
ENSG00000000419                              0.000036   
ENSG00000000457                              0.000692   
ENSG00000000460                              0.001713   
ENSG00000000938                              0.000446   
ENSG00000000971                              0.000310   

                 N49-Diagnoses_main_ICD10_N49_Inflammatory_disorders_of_male_genital_organs_not_elsewhere_classified  \
gene_id                                                                                                                
ENSG00000000419                                       1.020227e-10                                                     
ENSG00000000457                                       8.216060e-11                                                     
ENSG00000000460                                       1.119010e-10                                                     
ENSG00000000938                                       3.573442e-10                                                     
ENSG00000000971                                       1.397300e-10                                                     

                 100011_raw-Iron  \
gene_id                            
ENSG00000000419         0.000020   
ENSG00000000457         0.000089   
ENSG00000000460         0.000016   
ENSG00000000938         0.000006   
ENSG00000000971         0.000182   

                 5221-Index_of_best_refractometry_result_right  \
gene_id                                                          
ENSG00000000419                                       0.001497   
ENSG00000000457                                       0.004363   
ENSG00000000460                                       0.003566   
ENSG00000000938                                       0.004429   
ENSG00000000971                                       0.003160   

                 20003_1141150624-Treatmentmedication_code_zomig_25mg_tablet  \
gene_id                                                                        
ENSG00000000419                                       3.426770e-11             
ENSG00000000457                                       1.342669e-10             
ENSG00000000460                                       1.533228e-10             
ENSG00000000938                                       2.588022e-08             
ENSG00000000971                                       3.940395e-11             

                 S69-Diagnoses_main_ICD10_S69_Other_and_unspecified_injuries_of_wrist_and_hand  \
gene_id                                                                                          
ENSG00000000419                                       1.312300e-10                               
ENSG00000000457                                       4.223476e-10                               
ENSG00000000460                                       3.712300e-10                               
ENSG00000000938                                       3.461500e-10                               
ENSG00000000971                                       1.785600e-10                               

                 20024_1136-Job_code_deduced_Information_and_communication_technology_managers  \
gene_id                                                                                          
ENSG00000000419                                       1.508600e-09                               
ENSG00000000457                   

In [11]:
assert fastenloc_gene_associations.min().min() >= 0
assert fastenloc_gene_associations.max().max() <= 3

# S-MultiXcan

## Counts of significant associations

In [12]:
smultixcan_gene_associations.shape

(22515, 4091)

In [13]:
all_pvalues = pd.Series(smultixcan_gene_associations.values.flatten()).dropna()
display(all_pvalues.shape)

(91055810,)

In [14]:
all_pvalues.describe()

count     9.105581e+07
mean      4.839226e-01
std       2.934763e-01
min      9.999889e-321
25%       2.266266e-01
50%       4.794625e-01
75%       7.378647e-01
max       1.000000e+00
dtype: float64

In [15]:
(-np.log10(all_pvalues)).describe()

count    9.105581e+07
mean     4.990325e-01
std      1.258476e+00
min     -0.000000e+00
25%      1.320233e-01
50%      3.192453e-01
75%      6.446892e-01
max      3.200000e+02
dtype: float64

In [16]:
PVALUE_THRESHOLD = (0.05 / all_pvalues.shape[0])
display(PVALUE_THRESHOLD)

5.491137797796759e-10

In [17]:
PVALUE_THRESHOLD = 5.49e-10

In [18]:
hits = (all_pvalues < PVALUE_THRESHOLD).sum()
display(hits)
display((hits / all_pvalues.shape[0]) * 100)

72994

0.08016402248247531

In [19]:
hits = (all_pvalues < 0.01).sum()
display(hits)
display((hits / all_pvalues.shape[0]) * 100)

1809587

1.9873383148203285

# fastENLOC

## Numbers of significant genes also found by fastENLOC

In [20]:
common_genes = fastenloc_gene_associations.index.intersection(smultixcan_gene_associations.index)
display(common_genes)

Index(['ENSG00000000419', 'ENSG00000000457', 'ENSG00000000460',
       'ENSG00000000938', 'ENSG00000000971', 'ENSG00000001036',
       'ENSG00000001084', 'ENSG00000001167', 'ENSG00000001460',
       'ENSG00000001461',
       ...
       'ENSG00000284240', 'ENSG00000284308', 'ENSG00000284395',
       'ENSG00000284413', 'ENSG00000284418', 'ENSG00000284430',
       'ENSG00000284452', 'ENSG00000284513', 'ENSG00000284526',
       'ENSG00000284552'],
      dtype='object', length=22427)

In [21]:
fastenloc_matrix_sorted = fastenloc_gene_associations.loc[common_genes, smultixcan_gene_associations.columns]
display(fastenloc_matrix_sorted.shape)
display(fastenloc_matrix_sorted.head())

(22427, 4091)

20096_1-Size_of_red_wine_glass_drunk_small_125ml  \
ENSG00000000419                                      2.173200e-07   
ENSG00000000457                                      1.343900e-06   
ENSG00000000460                                      8.757750e-06   
ENSG00000000938                                      7.197000e-08   
ENSG00000000971                                      2.720000e-07   

                 2345-Ever_had_bowel_cancer_screening  \
ENSG00000000419                              0.000036   
ENSG00000000457                              0.000692   
ENSG00000000460                              0.001713   
ENSG00000000938                              0.000446   
ENSG00000000971                              0.000310   

                 N49-Diagnoses_main_ICD10_N49_Inflammatory_disorders_of_male_genital_organs_not_elsewhere_classified  \
ENSG00000000419                                       1.020227e-10                                                     
ENSG00000000457                                       8.216060e-11                                                     
ENSG00000000460                                       1.119010e-10                                                     
ENSG00000000938                                       3.573442e-10                                                     
ENSG00000000971                                       1.397300e-10                                                     

                 100011_raw-Iron  \
ENSG00000000419         0.000020   
ENSG00000000457         0.000089   
ENSG00000000460         0.000016   
ENSG00000000938         0.000006   
ENSG00000000971         0.000182   

                 5221-Index_of_best_refractometry_result_right  \
ENSG00000000419                                       0.001497   
ENSG00000000457                                       0.004363   
ENSG00000000460                                       0.003566   
ENSG00000000938                                       0.004429   
ENSG00000000971                                       0.003160   

                 20003_1141150624-Treatmentmedication_code_zomig_25mg_tablet  \
ENSG00000000419                                       3.426770e-11             
ENSG00000000457                                       1.342669e-10             
ENSG00000000460                                       1.533228e-10             
ENSG00000000938                                       2.588022e-08             
ENSG00000000971                                       3.940395e-11             

                 S69-Diagnoses_main_ICD10_S69_Other_and_unspecified_injuries_of_wrist_and_hand  \
ENSG00000000419                                       1.312300e-10                               
ENSG00000000457                                       4.223476e-10                               
ENSG00000000460                                       3.712300e-10                               
ENSG00000000938                                       3.461500e-10                               
ENSG00000000971                                       1.785600e-10                               

                 20024_1136-Job_code_deduced_Information_and_communication_technology_managers  \
ENSG00000000419                                       1.508600e-09                               
ENSG00000000457                                       2.036232e-08                               
ENSG00000000460                                       6.505700e-09                               
ENSG00000000938                                       9.110200e-09                               
ENSG00000000971                                       2.329954e-10                               

                 20002_1385-Noncancer_illness_code_selfreported_allergy_or_anaphylactic_reaction_to_food  \
ENSG00000000419                                       1.114500e-05                                         
ENSG00000000457                                  

In [22]:
multixcan_matrix_sorted = smultixcan_gene_associations.loc[common_genes]
display(multixcan_matrix_sorted.shape)
display(multixcan_matrix_sorted.head())

(22427, 4091)

20096_1-Size_of_red_wine_glass_drunk_small_125ml  \
ENSG00000000419                                          0.865429   
ENSG00000000457                                          0.174192   
ENSG00000000460                                          0.879969   
ENSG00000000938                                          0.192670   
ENSG00000000971                                          0.180632   

                 2345-Ever_had_bowel_cancer_screening  \
ENSG00000000419                              0.918314   
ENSG00000000457                              0.064765   
ENSG00000000460                              0.240715   
ENSG00000000938                              0.400054   
ENSG00000000971                              0.793060   

                 N49-Diagnoses_main_ICD10_N49_Inflammatory_disorders_of_male_genital_organs_not_elsewhere_classified  \
ENSG00000000419                                           0.810683                                                     
ENSG00000000457                                           0.889194                                                     
ENSG00000000460                                           0.238228                                                     
ENSG00000000938                                           0.114353                                                     
ENSG00000000971                                           0.490585                                                     

                 100011_raw-Iron  \
ENSG00000000419         0.374671   
ENSG00000000457         0.896938   
ENSG00000000460         0.567555   
ENSG00000000938         0.470700   
ENSG00000000971         0.088752   

                 5221-Index_of_best_refractometry_result_right  \
ENSG00000000419                                       0.189032   
ENSG00000000457                                       0.448596   
ENSG00000000460                                       0.921320   
ENSG00000000938                                       0.889202   
ENSG00000000971                                       0.744531   

                 20003_1141150624-Treatmentmedication_code_zomig_25mg_tablet  \
ENSG00000000419                                           0.140981             
ENSG00000000457                                           0.269602             
ENSG00000000460                                           0.825036             
ENSG00000000938                                           0.000011             
ENSG00000000971                                           0.949639             

                 S69-Diagnoses_main_ICD10_S69_Other_and_unspecified_injuries_of_wrist_and_hand  \
ENSG00000000419                                           0.467741                               
ENSG00000000457                                           0.540261                               
ENSG00000000460                                           0.782230                               
ENSG00000000938                                           0.899764                               
ENSG00000000971                                           0.253817                               

                 20024_1136-Job_code_deduced_Information_and_communication_technology_managers  \
ENSG00000000419                                           0.129427                               
ENSG00000000457                                           0.068405                               
ENSG00000000460                                           0.644525                               
ENSG00000000938                                           0.212352                               
ENSG00000000971                                           0.377408                               

                 20002_1385-Noncancer_illness_code_selfreported_allergy_or_anaphylactic_reaction_to_food  \
ENSG00000000419                                           0.193680                                         
ENSG00000000457                                  

In [23]:
assert fastenloc_matrix_sorted.shape == multixcan_matrix_sorted.shape

### Using stringent pvalue threshold

In [24]:
display(PVALUE_THRESHOLD)

5.49e-10

In [25]:
multixcan_signif = (multixcan_matrix_sorted < PVALUE_THRESHOLD)
display(multixcan_signif.shape)
display(multixcan_signif.head())

(22427, 4091)

20096_1-Size_of_red_wine_glass_drunk_small_125ml  \
ENSG00000000419                                             False   
ENSG00000000457                                             False   
ENSG00000000460                                             False   
ENSG00000000938                                             False   
ENSG00000000971                                             False   

                 2345-Ever_had_bowel_cancer_screening  \
ENSG00000000419                                 False   
ENSG00000000457                                 False   
ENSG00000000460                                 False   
ENSG00000000938                                 False   
ENSG00000000971                                 False   

                 N49-Diagnoses_main_ICD10_N49_Inflammatory_disorders_of_male_genital_organs_not_elsewhere_classified  \
ENSG00000000419                                              False                                                     
ENSG00000000457                                              False                                                     
ENSG00000000460                                              False                                                     
ENSG00000000938                                              False                                                     
ENSG00000000971                                              False                                                     

                 100011_raw-Iron  \
ENSG00000000419            False   
ENSG00000000457            False   
ENSG00000000460            False   
ENSG00000000938            False   
ENSG00000000971            False   

                 5221-Index_of_best_refractometry_result_right  \
ENSG00000000419                                          False   
ENSG00000000457                                          False   
ENSG00000000460                                          False   
ENSG00000000938                                          False   
ENSG00000000971                                          False   

                 20003_1141150624-Treatmentmedication_code_zomig_25mg_tablet  \
ENSG00000000419                                              False             
ENSG00000000457                                              False             
ENSG00000000460                                              False             
ENSG00000000938                                              False             
ENSG00000000971                                              False             

                 S69-Diagnoses_main_ICD10_S69_Other_and_unspecified_injuries_of_wrist_and_hand  \
ENSG00000000419                                              False                               
ENSG00000000457                                              False                               
ENSG00000000460                                              False                               
ENSG00000000938                                              False                               
ENSG00000000971                                              False                               

                 20024_1136-Job_code_deduced_Information_and_communication_technology_managers  \
ENSG00000000419                                              False                               
ENSG00000000457                                              False                               
ENSG00000000460                                              False                               
ENSG00000000938                                              False                               
ENSG00000000971                                              False                               

                 20002_1385-Noncancer_illness_code_selfreported_allergy_or_anaphylactic_reaction_to_food  \
ENSG00000000419                                              False                                         
ENSG00000000457                                  

In [26]:
fastenloc_signif = (fastenloc_matrix_sorted > 0.1)
display(fastenloc_signif.shape)
display(fastenloc_signif.head())

(22427, 4091)

20096_1-Size_of_red_wine_glass_drunk_small_125ml  \
ENSG00000000419                                             False   
ENSG00000000457                                             False   
ENSG00000000460                                             False   
ENSG00000000938                                             False   
ENSG00000000971                                             False   

                 2345-Ever_had_bowel_cancer_screening  \
ENSG00000000419                                 False   
ENSG00000000457                                 False   
ENSG00000000460                                 False   
ENSG00000000938                                 False   
ENSG00000000971                                 False   

                 N49-Diagnoses_main_ICD10_N49_Inflammatory_disorders_of_male_genital_organs_not_elsewhere_classified  \
ENSG00000000419                                              False                                                     
ENSG00000000457                                              False                                                     
ENSG00000000460                                              False                                                     
ENSG00000000938                                              False                                                     
ENSG00000000971                                              False                                                     

                 100011_raw-Iron  \
ENSG00000000419            False   
ENSG00000000457            False   
ENSG00000000460            False   
ENSG00000000938            False   
ENSG00000000971            False   

                 5221-Index_of_best_refractometry_result_right  \
ENSG00000000419                                          False   
ENSG00000000457                                          False   
ENSG00000000460                                          False   
ENSG00000000938                                          False   
ENSG00000000971                                          False   

                 20003_1141150624-Treatmentmedication_code_zomig_25mg_tablet  \
ENSG00000000419                                              False             
ENSG00000000457                                              False             
ENSG00000000460                                              False             
ENSG00000000938                                              False             
ENSG00000000971                                              False             

                 S69-Diagnoses_main_ICD10_S69_Other_and_unspecified_injuries_of_wrist_and_hand  \
ENSG00000000419                                              False                               
ENSG00000000457                                              False                               
ENSG00000000460                                              False                               
ENSG00000000938                                              False                               
ENSG00000000971                                              False                               

                 20024_1136-Job_code_deduced_Information_and_communication_technology_managers  \
ENSG00000000419                                              False                               
ENSG00000000457                                              False                               
ENSG00000000460                                              False                               
ENSG00000000938                                              False                               
ENSG00000000971                                              False                               

                 20002_1385-Noncancer_illness_code_selfreported_allergy_or_anaphylactic_reaction_to_food  \
ENSG00000000419                                              False                                         
ENSG00000000457                                  

In [27]:
multixcan_hits = multixcan_signif.sum().sum()
display(multixcan_hits)
display(multixcan_hits / multixcan_signif.size)

72990

0.000795541245816283

In [28]:
fastenloc_hits = fastenloc_signif.sum().sum()
display(fastenloc_hits)
display(fastenloc_hits / fastenloc_signif.size)

292506

0.0031881160110801164

In [29]:
fastenloc_and_multixcan_signif = (multixcan_signif & fastenloc_signif)
assert fastenloc_and_multixcan_signif.shape == multixcan_signif.shape == fastenloc_signif.shape

In [30]:
fe_mu_sum = fastenloc_and_multixcan_signif.sum().sum()
display(fe_mu_sum)
_perc = fe_mu_sum / multixcan_hits
display(_perc)
display(f'{(_perc * 100):.2f}%')

22473

0.30789149198520344

'30.79%'

### Conclusions

72,990 significant associations with S-MultiXcan (after taking the set of genes in common with fastENLOC, otherwise it's 72,994).

From these, 22,473 associations are also colocalized (30.79%)

# Export significant associations as a table

### S-MultiXcan

In [31]:
smultixcan_results = multixcan_matrix_sorted.unstack()

In [32]:
smultixcan_results.shape

(91748857,)

In [33]:
smultixcan_results.head()

20096_1-Size_of_red_wine_glass_drunk_small_125ml  ENSG00000000419    0.865429
                                                  ENSG00000000457    0.174192
                                                  ENSG00000000460    0.879969
                                                  ENSG00000000938    0.192670
                                                  ENSG00000000971    0.180632
dtype: float64

In [34]:
smultixcan_results.isna().sum()

1049005

In [35]:
smultixcan_results = smultixcan_results.dropna()

In [36]:
smultixcan_results.shape

(90699852,)

In [37]:
assert smultixcan_results.isna().sum() == 0

In [38]:
(smultixcan_results < PVALUE_THRESHOLD).sum()

72990

### fastENLOC

In [39]:
fastenloc_results = fastenloc_matrix_sorted.unstack()

In [40]:
fastenloc_results.shape

(91748857,)

In [41]:
fastenloc_results.head()

20096_1-Size_of_red_wine_glass_drunk_small_125ml  ENSG00000000419    2.173200e-07
                                                  ENSG00000000457    1.343900e-06
                                                  ENSG00000000460    8.757750e-06
                                                  ENSG00000000938    7.197000e-08
                                                  ENSG00000000971    2.720000e-07
dtype: float64

In [42]:
fastenloc_results.isna().sum()

1554545

In [43]:
fastenloc_results = fastenloc_results.dropna()

In [44]:
fastenloc_results.shape

(90194312,)

In [45]:
final_results = pd.DataFrame(index=smultixcan_results.index.copy(), columns=['smultixcan_pval', 'fastenloc_rcp'])

In [46]:
final_results = final_results.assign(smultixcan_pval=smultixcan_results, fastenloc_rcp=fastenloc_results)

In [47]:
final_results.shape

(90699852, 2)

In [48]:
final_results.head()

smultixcan_pval  \
20096_1-Size_of_red_wine_glass_drunk_small_125ml ENSG00000000419         0.865429   
                                                 ENSG00000000457         0.174192   
                                                 ENSG00000000460         0.879969   
                                                 ENSG00000000938         0.192670   
                                                 ENSG00000000971         0.180632   

                                                                  fastenloc_rcp  
20096_1-Size_of_red_wine_glass_drunk_small_125ml ENSG00000000419   2.173200e-07  
                                                 ENSG00000000457   1.343900e-06  
                                                 ENSG00000000460   8.757750e-06  
                                                 ENSG00000000938   7.197000e-08  
                                                 ENSG00000000971   2.720000e-07

In [49]:
final_results.isna().sum()

smultixcan_pval          0
fastenloc_rcp      1529981
dtype: int64

In [50]:
final_results = final_results.dropna(how='any')

In [51]:
final_results.shape

(89169871, 2)

In [52]:
PVALUE_THRESHOLD

5.49e-10

In [53]:
publishable_final_results = final_results[(final_results['smultixcan_pval'] < PVALUE_THRESHOLD) & (final_results['fastenloc_rcp'] > 0.1)]

In [54]:
publishable_final_results.index.rename(('trait', 'gene_id'), inplace=True)

In [55]:
publishable_final_results.shape

(22473, 2)

In [56]:
publishable_final_results.head()

smultixcan_pval  \
trait                                       gene_id                            
806-Job_involves_mainly_walking_or_standing ENSG00000067560     2.318351e-12   
                                            ENSG00000145022     4.778520e-12   
                                            ENSG00000145029     2.400440e-11   
                                            ENSG00000164061     1.577103e-11   
                                            ENSG00000164062     4.626823e-13   

                                                             fastenloc_rcp  
trait                                       gene_id                         
806-Job_involves_mainly_walking_or_standing ENSG00000067560       0.254800  
                                            ENSG00000145022       0.363500  
                                            ENSG00000145029       0.307000  
                                            ENSG00000164061       0.233301  
                                            ENSG00000164062       0.919720

In [57]:
metadata.GENES_MAPPINGS.head()

gene gene_name       gene_type          gene_id    band
0  ENSG00000000457.13     SCYL3  protein_coding  ENSG00000000457  1q24.2
1  ENSG00000000460.16  C1orf112  protein_coding  ENSG00000000460  1q24.2
2  ENSG00000000938.12       FGR  protein_coding  ENSG00000000938  1p35.3
3  ENSG00000000971.15       CFH  protein_coding  ENSG00000000971  1q31.3
4  ENSG00000001036.13     FUCA2  protein_coding  ENSG00000001036  6q24.2

In [58]:
_tmp = pd.merge(
    publishable_final_results.reset_index(),
    metadata.GENES_MAPPINGS[['gene_id', 'gene_name', 'band', 'gene_type']],
    on='gene_id',
    how='left'
).rename(columns={'band': 'gene_band'})

In [59]:
_tmp.head()

trait          gene_id  \
0  806-Job_involves_mainly_walking_or_standing  ENSG00000067560   
1  806-Job_involves_mainly_walking_or_standing  ENSG00000145022   
2  806-Job_involves_mainly_walking_or_standing  ENSG00000145029   
3  806-Job_involves_mainly_walking_or_standing  ENSG00000164061   
4  806-Job_involves_mainly_walking_or_standing  ENSG00000164062   

   smultixcan_pval  fastenloc_rcp gene_name gene_band       gene_type  
0     2.318351e-12       0.254800      RHOA   3p21.31  protein_coding  
1     4.778520e-12       0.363500      TCTA   3p21.31  protein_coding  
2     2.400440e-11       0.307000     NICN1   3p21.31  protein_coding  
3     1.577103e-11       0.233301       BSN   3p21.31  protein_coding  
4     4.626823e-13       0.919720      APEH   3p21.31  protein_coding

In [60]:
publishable_final_results = _tmp.set_index(['trait', 'gene_id'])[['gene_name', 'gene_band', 'gene_type', 'smultixcan_pval', 'fastenloc_rcp']]

In [61]:
publishable_final_results = publishable_final_results.sort_index()

In [62]:
assert publishable_final_results.index.is_unique

In [63]:
publishable_final_results.isna().sum()

gene_name          0
gene_band          0
gene_type          0
smultixcan_pval    0
fastenloc_rcp      0
dtype: int64

In [64]:
publishable_final_results[publishable_final_results['gene_name'].isna()]

Empty DataFrame
Columns: [gene_name, gene_band, gene_type, smultixcan_pval, fastenloc_rcp]
Index: []

In [65]:
publishable_final_results.head()

gene_name gene_band  \
trait                                gene_id                               
102_raw-Pulse_rate_automated_reading ENSG00000031003    FAM13B    5q31.2   
                                     ENSG00000064989    CALCRL    2q32.1   
                                     ENSG00000072195      SPEG      2q35   
                                     ENSG00000075420    FNDC3B   3q26.31   
                                     ENSG00000075702     WDR62  19q13.12   

                                                           gene_type  \
trait                                gene_id                           
102_raw-Pulse_rate_automated_reading ENSG00000031003  protein_coding   
                                     ENSG00000064989  protein_coding   
                                     ENSG00000072195  protein_coding   
                                     ENSG00000075420  protein_coding   
                                     ENSG00000075702  protein_coding   

                                                      smultixcan_pval  \
trait                                gene_id                            
102_raw-Pulse_rate_automated_reading ENSG00000031003     9.316678e-29   
                                     ENSG00000064989     1.553027e-22   
                                     ENSG00000072195     1.707185e-38   
                                     ENSG00000075420     2.695478e-24   
                                     ENSG00000075702     8.858931e-11   

                                                      fastenloc_rcp  
trait                                gene_id                         
102_raw-Pulse_rate_automated_reading ENSG00000031003       0.868600  
                                     ENSG00000064989       0.863200  
                                     ENSG00000072195       0.877000  
                                     ENSG00000075420       0.841100  
                                     ENSG00000075702       1.257727

In [66]:
publishable_final_results.shape

(22473, 5)

In [67]:
# some testing
_result = publishable_final_results.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['VIP'])]
assert _result['smultixcan_pval'] == 1.8124965657173678e-17
assert _result['fastenloc_rcp'] == 0.30421

In [68]:
# some testing
_result = publishable_final_results.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['RP11-220I1.5'])]
assert _result['smultixcan_pval'] == 6.426619359444203e-11
assert _result['fastenloc_rcp'] == 0.2150000007971

In [69]:
# for publication (tsv.gz)
output_folder = os.path.join(conf.DELIVERABLES_DIR, 'supp_tables')
os.makedirs(output_folder, exist_ok=True)
output_file = os.path.join(output_folder, 'suppl_table_S1-significant_gene_trait_associations.tsv.gz')
display(output_file)

publishable_final_results.to_csv(output_file, sep='\t', float_format='%.4e')

'/mnt/phenomexcan_base/deliverables/supp_tables/suppl_table_S1-significant_gene_trait_associations.tsv.gz'

In [70]:
# some testing
_tmp = pd.read_csv(output_file, sep='\t')

In [71]:
assert publishable_final_results.shape == _tmp.set_index(['trait', 'gene_id']).shape

In [72]:
_tmp.shape

(22473, 7)

In [73]:
_tmp.head()

trait          gene_id gene_name gene_band  \
0  102_raw-Pulse_rate_automated_reading  ENSG00000031003    FAM13B    5q31.2   
1  102_raw-Pulse_rate_automated_reading  ENSG00000064989    CALCRL    2q32.1   
2  102_raw-Pulse_rate_automated_reading  ENSG00000072195      SPEG      2q35   
3  102_raw-Pulse_rate_automated_reading  ENSG00000075420    FNDC3B   3q26.31   
4  102_raw-Pulse_rate_automated_reading  ENSG00000075702     WDR62  19q13.12   

        gene_type  smultixcan_pval  fastenloc_rcp  
0  protein_coding     9.316700e-29         0.8686  
1  protein_coding     1.553000e-22         0.8632  
2  protein_coding     1.707200e-38         0.8770  
3  protein_coding     2.695500e-24         0.8411  
4  protein_coding     8.858900e-11         1.2577

In [74]:
_tmp = _tmp.set_index(['trait', 'gene_id'])

In [75]:
_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['VIP'])]
assert _result['smultixcan_pval'] == 1.8125e-17
assert _result['fastenloc_rcp'] == 0.30421

_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['RP11-220I1.5'])]
assert _result['smultixcan_pval'] == 6.4266e-11
assert _result['fastenloc_rcp'] == 0.2150

In [76]:
assert np.allclose(
    _tmp[['smultixcan_pval', 'fastenloc_rcp']].values,
    publishable_final_results[['smultixcan_pval', 'fastenloc_rcp']].values,
    atol=1e-320,
    rtol=1e-4
)

In [77]:
# for publication (xlsx)
output_folder = os.path.join(conf.DELIVERABLES_DIR, 'supp_tables')
os.makedirs(output_folder, exist_ok=True)
output_file = os.path.join(output_folder, 'suppl_table_S1-significant_gene_trait_associations.xlsx')
display(output_file)

publishable_final_results.reset_index().to_excel(output_file, index=False, float_format='%.4e')

'/mnt/phenomexcan_base/deliverables/supp_tables/suppl_table_S1-significant_gene_trait_associations.xlsx'

In [78]:
# some testing
_tmp = pd.read_excel(output_file)

In [79]:
assert publishable_final_results.shape == _tmp.set_index(['trait', 'gene_id']).shape

In [80]:
_tmp.shape

(22473, 7)

In [81]:
_tmp.head()

trait          gene_id gene_name gene_band  \
0  102_raw-Pulse_rate_automated_reading  ENSG00000031003    FAM13B    5q31.2   
1  102_raw-Pulse_rate_automated_reading  ENSG00000064989    CALCRL    2q32.1   
2  102_raw-Pulse_rate_automated_reading  ENSG00000072195      SPEG      2q35   
3  102_raw-Pulse_rate_automated_reading  ENSG00000075420    FNDC3B   3q26.31   
4  102_raw-Pulse_rate_automated_reading  ENSG00000075702     WDR62  19q13.12   

        gene_type  smultixcan_pval  fastenloc_rcp  
0  protein_coding     9.316700e-29         0.8686  
1  protein_coding     1.553000e-22         0.8632  
2  protein_coding     1.707200e-38         0.8770  
3  protein_coding     2.695500e-24         0.8411  
4  protein_coding     8.858900e-11         1.2577

In [82]:
_tmp = _tmp.set_index(['trait', 'gene_id'])

In [83]:
_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['VIP'])]
assert _result['smultixcan_pval'] == 1.8125e-17
assert _result['fastenloc_rcp'] == 0.30421

_result = _tmp.loc[('1180-Morningevening_person_chronotype', metadata.GENE_NAME_TO_ID_MAP['RP11-220I1.5'])]
assert _result['smultixcan_pval'] == 6.4266e-11
assert _result['fastenloc_rcp'] == 0.2150

In [84]:
assert np.allclose(
    _tmp[['smultixcan_pval', 'fastenloc_rcp']].values,
    publishable_final_results[['smultixcan_pval', 'fastenloc_rcp']].values,
    atol=1e-320,
    rtol=1e-4
)

# Code below needs to be moved out from here

# QQ plots

In [85]:
import plots
from clustering.biclustering.analysis import Trait

ModuleNotFoundError: No module named 'plots'

# Cholesterol

In [ ]:
spredixcan_genes_associations.columns[spredixcan_genes_associations.columns.str.lower().str.contains('cholesterol')]

In [ ]:
pheno_plain_name = '20002_1473-Noncancer_illness_code_selfreported_high_cholesterol'

In [ ]:
pheno_pvalues = pd.Series(spredixcan_genes_associations[pheno_plain_name].values.flatten())
display(pheno_pvalues.shape)

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=pheno_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = pheno_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})
display(df.shape)

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])
ax.set_title(Trait(pheno_plain_name).get_pretty_name())

# Height

In [ ]:
spredixcan_genes_associations.columns[spredixcan_genes_associations.columns.str.lower().str.contains('height')]

In [ ]:
pheno_plain_name = '50_raw-Standing_height'

In [ ]:
pheno_pvalues = pd.Series(spredixcan_genes_associations[pheno_plain_name].values.flatten())
display(pheno_pvalues.shape)

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=pheno_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = pheno_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})
display(df.shape)

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])
ax.set_title(Trait(pheno_plain_name).get_pretty_name())

# Schizophrenia

In [ ]:
spredixcan_genes_associations.columns[spredixcan_genes_associations.columns.str.lower().str.contains('schizophrenia')]

In [ ]:
pheno_plain_name = '20002_1289-Noncancer_illness_code_selfreported_schizophrenia'

In [ ]:
pheno_pvalues = pd.Series(spredixcan_genes_associations[pheno_plain_name].values.flatten())
display(pheno_pvalues.shape)

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=pheno_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = pheno_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})
display(df.shape)

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])
ax.set_title(Trait(pheno_plain_name).get_pretty_name())

# All p-values

In [ ]:
threshold = 1e-30

unif_values = np.linspace(0.0001, 1, num=all_pvalues.shape[0])
unif_values[unif_values < threshold] = threshold

real_values = all_pvalues
real_values[real_values < threshold] = threshold

df = pd.DataFrame({'unif': unif_values, 'real': real_values})

In [ ]:
df_plot = df

ax = plots.qqplot(-np.log10(df_plot), 'unif', 'real', xlabel='Expected $-\log_{10}(p)$', ylabel='Observed $-\log_{10}(p)$', s=5, linewidth=0)
ax.set_xlim([0, 5])